File: src/lib/tauri/connections.ts
import { invoke } from "@tauri-apps/api/core";

import type { ConnectionProfile, SaveConnectionInput } from "@/types/connections";

const LIST_CONNECTIONS_COMMAND = "list_connections";
const CREATE_CONNECTION_COMMAND = "create_connection";
const UPDATE_CONNECTION_COMMAND = "update_connection";
const DELETE_CONNECTION_COMMAND = "delete_connection";

export function listConnections() {
  return invoke<ConnectionProfile[]>(LIST_CONNECTIONS_COMMAND);
}

export function createConnection(input: SaveConnectionInput) {
  return invoke<ConnectionProfile>(CREATE_CONNECTION_COMMAND, { input });
}

export function updateConnection(id: string, input: SaveConnectionInput) {
  return invoke<ConnectionProfile>(UPDATE_CONNECTION_COMMAND, { id, input });
}

export function deleteConnection(id: string) {
  return invoke<void>(DELETE_CONNECTION_COMMAND, { id });
}

File: src-tauri/src/commands/connections.rs
use tauri::State;

use crate::models::connections::{ConnectionProfileDto, SaveConnectionInput};
use crate::models::error::CommandError;
use crate::services::connections::ConnectionService;

#[tauri::command]
pub async fn list_connections(
    service: State<'_, ConnectionService>,
) -> Result<Vec<ConnectionProfileDto>, CommandError> {
    service
        .list()
        .await
        .map(|profiles| profiles.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}

#[tauri::command]
pub async fn create_connection(
    input: SaveConnectionInput,
    service: State<'_, ConnectionService>,
) -> Result<ConnectionProfileDto, CommandError> {
    let (draft, credential, _) = input.into_parts().map_err(CommandError::from)?;

    service
        .create(draft, credential)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn update_connection(
    id: String,
    input: SaveConnectionInput,
    service: State<'_, ConnectionService>,
) -> Result<ConnectionProfileDto, CommandError> {
    let (draft, credential, should_clear_credential) =
        input.into_parts().map_err(CommandError::from)?;

    service
        .update(id, draft, credential, should_clear_credential)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn delete_connection(
    id: String,
    service: State<'_, ConnectionService>,
) -> Result<(), CommandError> {
    service.delete(id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn reveal_connection_credential(
    id: String,
    service: State<'_, ConnectionService>,
) -> Result<Option<String>, CommandError> {
    service
        .reveal_credential(id)
        .await
        .map(|credential| credential.map(|secret| secret.take()))
        .map_err(Into::into)
}
